server/src/main/java/io/deephaven/server/barrage/BarrageStreamGenerator.java

```diff
@@ -699,14 +952,43 @@ private long appendModColumns(final View view, final long startRange, final long
             numRows = Math.max(numRows, myModOffsets.size());
 
             try {
-                final ChunkInputStreamGenerator.DrainableColumn drainableColumn =
-                        mcd.data.getInputStream(view.options(), myModOffsets);
-
-                drainableColumn.visitFieldNodes(fieldNodeListener);
-                drainableColumn.visitBuffers(bufferListener);
-
-                // See comment in appendAddColumns
-                addStream.accept(drainableColumn);
+                boolean columnWritten = false;
+                // iterate through each chunk and find the one that contains the requested data
+                for (int i = 0; i < mcd.data.size() && !columnWritten; ++i) {
+                    if (startRange <= mcd.modChunkRowSets[i].lastRowKey()) {
+                        final ChunkInputStreamGenerator generator = mcd.data.generators[i];
+
+                        final long shiftAmount = -mcd.modChunkRowSets[i].firstRowKey();
+
+                        // get an offset rowset for each row in this chunk
+                        try (final WritableRowSet adjustedOffsets = myModOffsets.intersect(mcd.modChunkRowSets[i])) {
+                            // normalize this to the chunk offsets
+                            adjustedOffsets.shiftInPlace(shiftAmount);
+
+                            final ChunkInputStreamGenerator.DrainableColumn drainableColumn =
+                                    generator.getInputStream(view.options(), adjustedOffsets);
+                            drainableColumn.visitFieldNodes(fieldNodeListener);
+                            drainableColumn.visitBuffers(bufferListener);
+
+                            // Add the drainable last as it is allowed to immediately close a row set the visitors need
+                            addStream.accept(drainableColumn);
+                        }
+                        // no need to test other chunks
+                        columnWritten = true;
+                    }
+                }
+                // must write column data, just write an empty column
+                if (!columnWritten) {
+                    try (final RowSet empty = RowSetFactory.empty()) {
+                        final ChunkInputStreamGenerator.DrainableColumn drainableColumn =
+                                mcd.data.generators[0].getInputStream(view.options(), empty);
+                        drainableColumn.visitFieldNodes(fieldNodeListener);
+                        drainableColumn.visitBuffers(bufferListener);
+
+                        // Add the drainable last as it is allowed to immediately close a row set the visitors need
+                        addStream.accept(drainableColumn);
+                    }
+                }
```

Comment:
similar comments for adds -- but, obviously, the `adjustedOffsets` is unique per column. For return value I would select the max rows written across all columns.